cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -595,30 +620,50 @@ func (m *stateMachine) setUnlockStateWithFallbackKey(partName string, unlockRes
 		return fmt.Errorf("inconsistent disk encryption status: previous access resulted in encrypted, but now is unencrypted from partition %s", partName)
 	}
 
-	// if isEncryptedDev hasn't been set on the state machine yet, then set that
-	// on the state machine before continuing - this is okay because we might
-	// not have been able to do anything with ubuntu-data if we couldn't mount
-	// ubuntu-boot, so this might be the first time we tried to unlock
-	// ubuntu-data and m.isEncryptedDev may have the default value of false
+	// There are a few cases where this could be the first time that we found a
+	// decrypted device in the UnlockResult, but m.isEncryptedDev is still
+	// false.
+	// - The first case is if we couldn't find ubuntu-boot at all, in which case
+	// we can't use the run object keys from there and instead need to directly
+	// fallback to trying the fallback object keys from ubuntu-seed
+	// - The second case is if we couldn't identify an ubuntu-data-enc or an
+	// ubuntu-data partition at all, we still could have an ubuntu-save-enc
+	// partition in which case we maybe could still have an encrypted disk that
+	// needs unlocking with the fallback object keys from ubuntu-seed
+	//
+	// As such, if m.isEncryptedDev is false, but unlockRes.IsDecryptedDevice is
+	// true, then it is safe to assign m.isEncryptedDev to true.
 	if !m.isEncryptedDev && unlockRes.IsDecryptedDevice {
 		m.isEncryptedDev = unlockRes.IsDecryptedDevice
 	}
 
-	// also make sure that if we previously saw a device that we see the same
-	// device again
-	if unlockRes.Device != "" && part.Device != "" && unlockRes.Device != part.Device {
-		return fmt.Errorf("inconsistent partitions found for %s: previously found %s but now found %s", partName, part.Device, unlockRes.Device)
+	// Also make sure that if we previously saw a device that we see the same
+	// device again.
+	if unlockRes.Device != "" && part.rawDevice != "" && unlockRes.Device != part.rawDevice {
+		return fmt.Errorf("inconsistent partitions found for %s: previously found %s but now found %s", partName, part.rawDevice, unlockRes.Device)
 	}
 
-	if unlockRes.Device != "" {
-		part.FindState = partitionFound
-		part.Device = unlockRes.Device
+	// Note that we don't do any consistency checking on part.decryptedRawDevice
+	// and unlockRes.DecryptedDevice because unlockRes.DecryptedDevice should
+	// only have been set if the decryption result succeeded, so
+	// we should never see multiple UnlockResult's with DecryptedDevice. This is
+	// because the first time that it succeeded with the run object key we would not have entered into
+	// the fallback case here.
+
+	// We do however do consistency checking on unlockRes to make sure the
+	// result makes sense.
+	if !unlockRes.IsDecryptedDevice && unlockRes.DecryptedDevice != "" && err != nil {
```

Comment:
the old code was using Device before,  I'm not quite sure what this check is checking right now, it's never failing in any test

Reply:
the test does exactly what the comment says it does? we are making sure that the unlockRes and unlockErr that were returned are logically consistent, otherwise a bug/logical error in secboot would have very confusing results here

Reply:
well regardless, I added now an equivalent but perhaps redundant/paranoid check around both fsDevice and partDevice agreeing with each other